indexer/src/store.rs

```diff
@@ -0,0 +1,1298 @@
+                let mut index_lock_hashes = HashSet::new();
+                index_lock_hashes.insert(lock_hash.to_owned());
+
+                let mut block = self
+                    .shared
+                    .store()
+                    .get_block(&index_state.block_hash)
+                    .expect("block exists");
+                // detach blocks until reach a block on main chain
+                self.commit_batch(|batch| {
+                    self.detach_block(batch, &index_lock_hashes, &block);
+                    while self
+                        .shared
+                        .store()
+                        .get_block_hash(block.header().number() - 1)
+                        != Some(block.header().parent_hash().to_owned())
+                    {
+                        block = self
+                            .shared
+                            .store()
+                            .get_block(block.header().parent_hash())
+                            .expect("block exists");
+                        self.detach_block(batch, &index_lock_hashes, &block);
+                    }
+                    let index_state = LockHashIndexState {
+                        block_number: block.header().number() - 1,
+                        block_hash: block.header().parent_hash().to_owned(),
+                    };
+                    batch.insert_lock_hash_index_state(lock_hash, &index_state);
+                });
+            });
+
+        // attach blocks until reach tip or batch limit
+        let mut lock_hash_index_states = self.get_lock_hash_index_states();
+        let min_block_number: BlockNumber = lock_hash_index_states
+            .values()
+            .min_by_key(|index_state| index_state.block_number)
+            .expect("none empty index states")
+            .block_number;
+
+        let (tip_number, tip_hash) = {
+            let tip_header = self
+                .shared
+                .store()
+                .get_tip_header()
+                .expect("tip header exists");
+            (tip_header.number(), tip_header.hash().to_owned())
+        };
+        self.commit_batch(|batch| {
+            (min_block_number + 1..=tip_number)
+                .take(BATCH_ATTACH_BLOCK_NUMS)
+                .for_each(|block_number| {
+                    let index_lock_hashes = lock_hash_index_states
+                        .iter()
+                        .filter(|(_, index_state)| index_state.block_number < block_number)
+                        .map(|(lock_hash, _)| lock_hash)
+                        .cloned()
+                        .collect();
+                    let block = self
+                        .shared
+                        .get_ancestor(&tip_hash, block_number)
+                        .and_then(|header| self.shared.store().get_block(&header.hash()))
+                        .expect("block exists");
+                    self.attach_block(batch, &index_lock_hashes, &block);
```

Comment:
Must check that the block is a child of the last indexed block. Because the index state may change between `detach` and `attach` when the thread is interrupted.

Reply:
won't fix this as same reason of https://github.com/nervosnetwork/ckb/pull/905#discussion_r294279649